File: src/array/design-hit-counter.ts
// DIFFICULTY: Medium
//
// Design a hit counter which counts the number of hits received in the past 5 minutes (i.e., the past 300 seconds).
//
// Your system should accept a timestamp parameter (in seconds granularity), and you may assume that calls are being
// made to the system in chronological order (i.e., timestamp is monotonically increasing). Several hits may arrive
// roughly at the same time.
//
// See {@link https://leetcode.com/problems/design-hit-counter/}
export { HitCounter };

// SOLUTION:
//
// To do this efficiently we'll have to use a circular array buffer.  This is the same technique used by time series
// databases.
//
// COMPLEXITY:
//
// Both methods run in O(1) time since we fix the array size at 300.
interface HitEvent {
  timestamp: number;
  hits: number;
}

class HitCounter {
  private readonly events: HitEvent[];

  constructor() {
    // Our granularity is seconds, so in theory, we only need to store only the last 300 seconds.  Keep track of both
    // hits by timestamp.
    this.events = [];

    // Don't use Array.fill() here; you'll fill the array with 300 copies of the same object.
    for (let i = 0; i < 300; i++) {
      this.events[i] = {
        timestamp: 0,
        hits: 0
      };
    }
  }

  hit(timestamp: number): void {
    // Find the event in our circular buffer that would correspond to this timestamp.
    const i = timestamp % 300;
    const event = this.events[i];

    // If we've already recorded hits at this timestamp, increment it.
    if (event.timestamp === timestamp) {
      event.hits++;
      return;
    }

    // Otherwise, we haven't, so set the hits to 1.
    event.timestamp = timestamp;
    event.hits = 1;
  }

  getHits(timestamp: number): number {
    let count = 0;

    for (const event of this.events) {
      if (timestamp - event.timestamp < 300) {
        count += event.hits;
      }
    }

    return count;
  }
}

File: src/leetcode/array/design_hit_counter.py
# DIFFICULTY: MEDIUM
# ------------------
#
# Design a hit counter which counts the number of hits received in the past 5 minutes (i.e., the past 300 seconds).
#
# Your system should accept a timestamp parameter (in seconds granularity), and you may assume that calls are being
# made to the system in chronological order (i.e., timestamp is monotonically increasing). Several hits may arrive
# roughly at the same time.
#
# See https://leetcode.com/problems/design-hit-counter
class HitCounter:
    def __init__(self) -> None:
        """
        SOLUTION
        --------

        To do this efficiently we'll have to use a circular array buffer.  This is the same technique used by time
        series databases.

        COMPLEXITY
        ----------

        Time complexity is O(1) for both methods since we fix the array size at 300.

        Space complexity is O(1) for both methods because we fix the array size at 300.
        """
        # Each event is a [timestamp, hit_count].  Since our granularity is in seconds, we only need to track the last
        # 300 seconds.
        self.timestamps = [0 for _ in range(300)]
        self.hits = [0 for _ in range(300)]

    def hit(self, timestamp: int) -> None:
        # Find the index of the event in our circular buffer that would correspond to this timestamp.
        i = timestamp % 300

        # If we've already recorded this timestamp, increment it.
        if self.timestamps[i] == timestamp:
            self.hits[i] += 1
        # If we haven't, set the hits to 1.
        else:
            self.timestamps[i] = timestamp
            self.hits[i] = 1

    def getHits(self, timestamp: int) -> int:
        count = 0

        for i, t in enumerate(self.timestamps):
            if timestamp - t < 300:
                count += self.hits[i]

        return count
